android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java

```diff
@@ -243,35 +241,169 @@ private void reevaluateContext() {
+        clearActiveKeys();
+        clearActiveVariants();
+
+
+        for (JsonElement experiment : configKeyStates.experiments) {
+
+            for (Map.Entry<String, JsonElement> elementEntry : experiment.getAsJsonObject().entrySet()) {
+
+                for (JsonElement jsonElement : elementEntry.getValue().getAsJsonArray()) {
 
-        evolvContext.set("keys.active", activeKeys, false);
-        evolvContext.set("variants.active", activeVariants, false);
+                    for (Map.Entry<String, JsonElement> activeKeyEntry : jsonElement.getAsJsonObject().entrySet()) {
+                        if(activeKeyEntry.getKey().startsWith("active_")){
+                            activeKeys.addProperty(activeKeyEntry.getKey(),activeKeyEntry.getValue().getAsString());
+                        }
+                    }
+                }
+            }
+            // TODO: 30.06.2021 add pruned
+        }
+
+        JsonObject newActiveKeys = activeKeys.deepCopy();
+        JsonObject newActiveVariants = activeVariants.deepCopy();
+
+        evolvContext.set("keys.active", newActiveKeys, false);
+        evolvContext.set("variants.active", newActiveVariants, false);
 
         // TODO: 01.06.2021 understand the data type  emit(,,?)
         waitForIt.emit(evolvContext, EFFECTIVE_GENOME_UPDATED, null);
 
-        // TODO: 01.06.2021 add
+        // TODO: 01.06.2021 add subscriptions
         reevaluatingContext = false;
     }
 
-    private Object generateEffectiveGenome(JsonElement expsKeyStates, JsonObject genomes) {
-        // TODO: 01.06.2021 implement
+    private void clearActiveKeys() {
+        for (String s : activeKeys.keySet()) activeKeys.remove(s);
+    }
+
+    private void clearActiveVariants() {
+        for (String s : activeVariants.keySet()) activeVariants.remove(s);
+    }
+
+    private Object generateEffectiveGenome(JsonArray expsKeyStates, JsonObject genomes) {
+
+        JsonObject effectiveGenome = new JsonObject();
+        JsonArray activeEids = new JsonArray();
+
+        for (JsonElement expKeyStates : expsKeyStates) {
+            //JsonObject active = expKeyStates.getAsJsonObject().get("active").getAsJsonObject();
+            //JsonObject eid = expKeyStates.getAsJsonObject().get("eid").getAsJsonObject();
+
+//            if (genomes.has("eid") && active) {
+//
+//            }
+
+        }
+
         return null;
     }
 
+    // TODO: 30.06.2021 need a unit test
     private void setActiveAndEntryKeyStates(int version,
                                             EvolvContext evolvContext,
                                             JsonObject config,
                                             JsonArray allocations,
                                             KeyStates configKeyStates) {
-        // TODO: 01.06.2021 implement
 
-        Object results = evaluatePredicates(version, evolvContext, config);
+        JsonArray results = evaluatePredicates(version, evolvContext, config);
```

Comment:
These types of changes need tests when you submit PRs for them

Reply:
done